### ISSUE TITLE
brew: Add explicit OS/arch dependencies to communicate compatibility gracefully

### ISSUE THREAD

**Issue by radeksimko** [goreleaser/goreleaser#2159]
**Describe the bug**

Users attempting to install a program via Homebrew formula generated by GoReleaser on unsupported OS/arch may receive confusing errors and be unable to `brew tap` the whole repo just because it contains invalid formulas with effectively missing `url` (after none of [these conditions](https://github.com/goreleaser/goreleaser/blob/41356ea84d734928493a7342d3c44e4b163237bc/internal/pipe/brew/template.go#L52-L89) are met).

**To Reproduce**

I don't have a full repro case (yet), but we maintain a generator that generates formulas in pretty much the same way as GoReleaser and people have reported failures on Apple Silicon https://github.com/hashicorp/homebrew-tap/issues/134 so I'm working with the assumption that this must affect GoReleaser too, it just probably isn't a common enough problem people run into (yet).

**Expected behavior**

User is presented with a clear error message that their OS/arch is not supported and (most importantly) they're able to still `brew tap` the tap even if it contains formulas which are not supported on their platform.

**Additional context**

I believe we could just set some default dependencies in formulas automatically as we know what's supported (OS/arch) based on the builds anyway.

For example
```ruby
depends_on arch: [:x86_64] if OS.mac?
```
or alternatively
```ruby
on_macos do
  depends_on arch: [:x86_64]
end
```

--- 

Apologies for the slightly vague description and lack of repro case - I'm just currently swamped with other unrelated work and wanted to create this issue to track the problem until I get to troubleshoot it properly and ideally send a patch.

**Comment by caarlos0** [comment]
sounds good, I researched a while ago how to do this and didn't found docs for it... glad to see its possible! 🙏

**Comment by caarlos0** [comment]
> Maybe it has been fixed?

no idea, need to test 🤔

**Comment by caarlos0** [comment]
check https://github.com/goreleaser/goreleaser/pull/2261

seems to be a bit better yes, per my tests at least...

if someone can help double-checking it would be amazing!

**Comment by caarlos0** [comment]
#2261 was partially reverted due to #2270

see Homebrew/discussions#1595

**Comment by caarlos0** [comment]
yeah, makes sense... I think the custom requirement is a good option... will look into how that can be made...

**Comment by ryancurrah** [comment]
Our company is starting to ship M1's to developers and when they are installing our internal tools through brew formula created by go releaser they are running into the same problem.

**Comment by ryancurrah** [comment]
That's something we discussed. Some tools use CGO and it isn't easy to support arm for them. So that's a challenge.

I know this is inherently not  goreleasers fault.

**Comment by caarlos0** [comment]
https://github.com/goreleaser/goreleaser/pull/2278 would probably be a better fix for it. I can rebase it at some point this week/weekend if you're willing to help testing it...

**Comment by ryancurrah** [comment]
For sure we can test it.

**Comment by caarlos0** [comment]
please check out  #2922

**Comment by caarlos0** [comment]
@ryancurrah were you able to give this a try?

**Comment by ryancurrah** [comment]
@caarlos0 I just did this morning and it does not address our issues. I posted a comment in the PR https://github.com/goreleaser/goreleaser/pull/2922#issuecomment-1050945941.
